pkg/apiserver/webhooks/handler.go

```diff
@@ -0,0 +1,198 @@
+	return token.NewGenerator(common.CloneTokenIssuer, key, 5*time.Minute)
+}
+
+func newAdmissionHandler(a Admitter) http.Handler {
+	return &admissionHandler{a: a}
+}
+
+func (h *admissionHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		klog.Errorf("contentType=%s, expect application/json", contentType)
+		return
+	}
+
+	klog.V(2).Info(fmt.Sprintf("handling request: %s", body))
+
+	// The AdmissionReview that was sent to the webhook
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		klog.Error(err)
+		responseAdmissionReview.Response = toAdmissionResponseError(err)
+	} else {
+		if requestedAdmissionReview.Request == nil {
+			responseAdmissionReview.Response = toAdmissionResponseError(fmt.Errorf("AdmissionReview.Request is nil"))
+		} else {
+			// pass to Admitter
+			responseAdmissionReview.Response = h.a.Admit(requestedAdmissionReview)
+		}
+	}
+
+	// Return the same UID
+	if requestedAdmissionReview.Request != nil {
+		responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+	}
+
+	klog.V(2).Info(fmt.Sprintf("sending response: %v", responseAdmissionReview.Response))
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		klog.Error(err)
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		klog.Error(err)
+	}
+}
+
+func toRejectedAdmissionResponse(causes []metav1.StatusCause) *v1beta1.AdmissionResponse {
+	globalMessage := ""
+	for _, cause := range causes {
+		globalMessage = fmt.Sprintf("%s %s", globalMessage, cause.Message)
+	}
+
+	return &v1beta1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: globalMessage,
+			Code:    http.StatusUnprocessableEntity,
+			Details: &metav1.StatusDetails{
+				Causes: causes,
+			},
+		},
+	}
+}
+
+func toAdmissionResponseError(err error) *v1beta1.AdmissionResponse {
+	return &v1beta1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+			Code:    http.StatusBadRequest,
+		},
+	}
+}
+
+func allowedAdmissionResponse() *admissionv1beta1.AdmissionResponse {
+	return &admissionv1beta1.AdmissionResponse{
+		Allowed: true,
+	}
+}
+
+func validateDataVolumeResource(ar v1beta1.AdmissionReview) error {
+	resource := metav1.GroupVersionResource{
+		Group:    cdicorev1alpha1.SchemeGroupVersion.Group,
+		Version:  cdicorev1alpha1.SchemeGroupVersion.Version,
+		Resource: "datavolumes",
+	}
+	if ar.Request.Resource != resource {
+		klog.Errorf("resource is %s but request is: %s", resource, ar.Request.Resource)
```

Comment:
These are GroupVersionResources, I don't think you can pass them as %s parameters. Should it be %v or resource.Resource as that is the name string.

Reply:
GroupVersionResources has a `String()` method